cli/src/commands/git/push.rs

```diff
@@ -279,6 +279,22 @@ pub fn cmd_git_push(
             &tx.base_workspace_helper().revset_parse_context(),
         )?)
         .range(&RevsetExpression::commits(new_heads));
+
+    // Only the roots themselves need to be checked if they're being pushed. If
+    // a root is already on the remote, then it is "grandfathered in" and
+    // shouldn't block any descendents from being pushed (unless the descendents
+    // are in private_roots()).
+    let private_commits_revset = revset_util::parse_private_roots_expression(
```

Comment:
What if the commits to be pushed are ancestors of the private roots?

Reply:
In that scenario, the commit in `private_roots()` is not part of commits_to_push and so the push is allowed.

My thought is: Whelp, that commit is already public, so clearly it shouldn’t be treated as though it’s private anymore.

Reply:
Then I believe the push will be allowed as long as the private root itself isn’t being pushed.

Reply:
I should make sure there’s an explicit tests for that too:

   o private
   |
   o public main*
   |
   o public main@origin

Pushing main should be allowed because the private commit isn’t being pushed.

Reply:
Added test `test_git_private_commits_descending_from_commits_pushed_do_not_block_pushing` for this.

Reply:
Oops, I misunderstood the definition. Sorry about the confusion.

BTW, if a typical use case is something like `private_roots = description("WIP")` (i.e. orthogonal to DAG), it's probably better to not define the roots, but just let user define which revision has to be private.

If the supposed use case is `private_roots = some_user_specified_branch_as_a_root_marker` (strictly defined based on DAG), any descendants of the roots shouldn't be allowed to push (no matter if the roots are already pushed.)

Reply:
I believe that's how it actually works. If a commit is in `private_roots()` but is already on the remote when you push, then its descendants aren't automatically blocked from pushing. Because you might have this:

```
o new stuff main*
|
o main@origin
|
= (400 elided commits)
|
o revert "WIP: building my feature"
|
o WIP: building my feature
|
~
```

That "WIP" commit is already public, so there's no sense blocking the push of _main_.

I thought "roots" was good in the name because of this:

```
o merge main*
|\
| o WIP: private changes
| |
o | public changes
|/
o main@origin
```

Even though the commit named "merge" isn't in `private_roots()`, it is implicitly blocked from pushing because doing so would require pushing the "WIP" commit as well.

The revset name can be changed though. If we prefer it to be something like `private_commits()`, that's no problem.

Reply:
I added a little more explanation to `revsets.md` and added a "Set of private commits" section to `config.md`.

Reply:
If it's named "roots", I would expect the rule is enforced to any `roots::` revisions, just like we do for `immutable_heads()` and `immutable() = ::immutable_heads()`.

Anyway, I don't think the global `private_roots()` concept is needed at this point. Some `jj git push`-specific config is probably good, and it can be extended to per-remote settings if needed. I'm not sure if `private_roots()` apply to other forges.

Reply:
_Ohhhh_, I think I understand your concerns now. Let me restate it to see if I'm understanding now.

First, the name `private_roots` implies too strongly that it would always make its descendants private, even if the root has already been pushed. I have no problem renaming it.

Second, you don't want private commits to be a "first-class feature" of jj, you'd rather it be a notion specific to the git backend. A backend like Google's doesn't really need this privacy concept since everything is published anyway.

If I made the following two changes, I think it addresses both concerns?

1. Remove the special "private" label and glyph I added to the built-in templates. The end-user can set this up if they want it.
2. Rename the config setting from `revset-aliases.private_roots()` to `git.private_commits()`.

Reply:
> First, the name `private_roots` implies too strongly that it would always make its descendants private, even if the root has already been pushed. I have no problem renaming it.
> 
> Second, you don't want private commits to be a "first-class feature" of jj, you'd rather it be a notion specific to the git backend. A backend like Google's doesn't really need this privacy concept since everything is published anyway.

Yes. Thanks for summarizing my thoughts.

I have no idea about Google backend, but I'm not sure if `gerrit send` should respect the "private" labeling for example. And we'll eventually need per-remote configuration because the user might want to exchange private commits between their local/personal clones.

>  1. Remove the special "private" label and glyph I added to the built-in templates. The end-user can set this up if they want it.
> 
>  2. Rename the config setting from `revset-aliases.private_roots()` to `git.private_commits()`.

Yes, that's the idea. The config name should be `git.private_commits` (no parentheses), btw. There might be a better name, but I don't have any suggestion.

This feature is similar to empty description check, so the default could be `private_commits = description(exact:"")`. I'm not sure if merging these is a good idea, and this can be addressed separately.

Reply:
I made these two changes:

- The setting is now named `git.private-commits` (the other options there used hyphens, not underscores, so I matched them)
- I dropped the commit that added glpyhs to the commit templater